### PR TITLE
[GLUTEN-6736][VL][1.2] Port #6741 #6783 #6658

### PR DESCRIPTION
**Pull request by wForget**
## What changes were proposed in this pull request?

Port #6741 #6783 #6658

### CODE REVIEW

File: cpp/velox/memory/VeloxMemoryManager.cc [hunk 1]
```diff
@@ -46,7 +46,7 @@ class ListenableArbitrator : public velox::memory::MemoryArbitrator {
   }
 
   uint64_t growCapacity(velox::memory::MemoryPool* pool, uint64_t targetBytes) override {
-    std::lock_guard<std::recursive_mutex> l(mutex_);
+    // std::lock_guard<std::recursive_mutex> l(mutex_);
```
Comment (wForget) [inline cpp/velox/memory/VeloxMemoryManager.cc:49]: @zhztheplayer Could you please help me figure out how to resolve this conflict? Can I just remove it?

Comment (zhztheplayer) [reply cpp/velox/memory/VeloxMemoryManager.cc:49]: I remember we don't need the locks like this one anymore. Could try removing.

Comment (wForget) [reply cpp/velox/memory/VeloxMemoryManager.cc:49]: Thank you, I will remove it.

File: cpp/velox/tests/MemoryManagerTest.cc [hunk 2]
```diff
@@ -129,4 +129,280 @@ TEST_F(MemoryManagerTest, memoryAllocatorWithBlockReservation) {
...
+
+  void set(AllocationListener* const delegate) {
+    if (delegate_ != nullptr) {
+      throw std::runtime_error("Invalid state");
+    }
+    delegate_ = delegate;
+  }
+
+  void allocationChanged(int64_t diff) override {
+    if (delegate_ == nullptr)  {
```
Comment (wForget) [inline cpp/velox/tests/MemoryManagerTest.cc:145]: This method may be called before delegate_ is assigned , so add this check.

Stack:
```
gluten::(anonymous namespace)::AllocationListenerWrapper::allocationChanged(long) (Unknown Source:0)
libvelox.so!gluten::ListenableArbitrator::growCapacity(facebook::velox::memory::MemoryPool*, unsigned long) (Unknown Source:0)
libvelox.so!facebook::velox::memory::MemoryManager::addRootPool(std::string const&, long, std::unique_ptr<facebook::velox::memory::MemoryReclaimer, std::default_delete<facebook::velox::memory::MemoryReclaimer> >) (Unknown Source:0)
libvelox.so!gluten::VeloxMemoryManager::VeloxMemoryManager(std::unique_ptr<gluten::AllocationListener, std::default_delete<gluten::AllocationListener> >) (Unknown Source:0)
std::thread::_State_impl<std::thread::_Invoker<std::tuple<gluten::MultiMemoryManagerTest_spill_Test::TestBody()::{lambda()#1}> > >::_M_run() (Unknown Source:0)
libgluten.so!execute_native_thread_routine (Unknown Source:0)
libpthread.so.0!start_thread (Unknown Source:0)
libc.so.6!clone (Unknown Source:0)
```

File: cpp/velox/tests/MemoryManagerTest.cc [hunk 3]
```diff
@@ -129,4 +129,280 @@ TEST_F(MemoryManagerTest, memoryAllocatorWithBlockReservation) {
...
+  }
+}
+} // namespace
+
+class MultiMemoryManagerTest : public ::testing::Test {
+ protected:
+  static void SetUpTestCase() {
+    std::unordered_map<std::string, std::string> conf = {
+        {kMemoryReservationBlockSize, std::to_string(kMemoryReservationBlockSizeDefault)},
+        {kVeloxMemInitCapacity, std::to_string(0)}};
```
Comment (wForget) [inline cpp/velox/tests/MemoryManagerTest.cc:340]: set veloxMemInitCapacity value to `0`.

File: cpp/velox/tests/MemoryManagerTest.cc [hunk 4]
```diff
@@ -129,4 +129,280 @@ TEST_F(MemoryManagerTest, memoryAllocatorWithBlockReservation) {
...
+      {
+        std::unique_lock<std::recursive_mutex> l(mutex);
+        vmms[i] = newVeloxMemoryManager(std::move(wrapper));
+        pool = vmms[i]->getLeafMemoryPool().get();
+        pool->setReclaimer(std::make_unique<MockMemoryReclaimer>(buffs[i], allocateSize));
+        listener->set(tmm.newListener(
+            [](uint64_t bytes) -> uint64_t { return 0; },
+            [i, &vmms, &mutex](uint64_t bytes) -> uint64_t {
+              std::unique_lock<std::recursive_mutex> l(mutex);
+              return vmms[i]->getMemoryManager()->shrinkPools(bytes);
```
Comment (wForget) [inline cpp/velox/tests/MemoryManagerTest.cc:382]: use `shrinkPools` method

**Comment by zhztheplayer** [comment]
Thank you for this effort. It's critical for the stability of release 1.2.

BTW, do you happen to know why `Velox backend Github Runner / run-tpc-test-ubuntu-oom` is failing? I will help investigate too asap.

**Comment by wForget** [comment]
> Thank you for this effort. It's critical for the stability of release 1.2.
> 
> BTW, do you happen to know why `Velox backend Github Runner / run-tpc-test-ubuntu-oom` is failing? I will help investigate too asap.

I haven't determined the cause of failure yet, I will try to verify separately commits to determine which change caused this failure.

**Comment by zhztheplayer** [comment]
Run Gluten Clickhouse CI

**Comment by zhztheplayer** [comment]
Hi @wForget , thank you for keeping working on this.

And I might forget that, have you picked https://github.com/apache/incubator-gluten/pull/7244? Which targeted to fix some hanging issues.

**Comment by wForget** [comment]
> Hi @wForget , thank you for keeping working on this.
> 
> And I might forget that, have you picked #7244? Which targeted to fix some hanging issues.

Thanks for your help, I will try pick #7244 later

**Comment by kecookier** [comment]
Could you kindly let me know if there are any issues that still need to be addressed with this PR?

**Comment by wForget** [comment]
> Could you please let me know if there are any issues with this PR that I should address?

There are still failed checks, do you have any ideas?

**Pull request closed without merging**